pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -139,16 +137,68 @@ const (
 	// So the meta data can be erased when pushing to envoy.
 	PilotMetaKey = "pilot_meta"
 
-	// TODO(yxue): separate h2c vs h2
-	H2Protocol = "h2"
-
 	// CanonicalHTTPSPort defines the standard port for HTTPS traffic. To avoid conflicts, http services
 	// are not allowed on this port.
 	CanonicalHTTPSPort = 443
+
+	// Alpn HTTP filter name which will override the ALPN for upstream TLS connection.
+	AlpnFilterName = "istio.alpn"
 )
 
+type FilterChainMatchOptions struct {
+	// Application protocols of the filter chain match
+	ApplicationProtocols []string
+	// Transport protocol of the filter chain match. "tls" or empty
+	TransportProtocol string
+	// Filter chain protocol. HTTP for rHTTP proxy and TCP for TCP proxy
+	Protocol plugin.ListenerProtocol
+}
+
 var (
-	applicationProtocols = []string{"http/1.0", "http/1.1"}
+	applicationProtocols = []string{"http/1.0", "http/1.1", "h2c"}
+
+	applicationProtocolsOverTLS = []string{"istio-http/1.0", "istio-http/1.1", "istio-h2"}
+
+	defaultApplicationProtocol = []string{"istio"}
+
+	// Double the number of filter chains. Half of filter chains are used as http filter chain and half of them are used as tcp proxy
+	// id in [0, len(allChains)/2) are configured as http filter chain, [(len(allChains)/2, len(allChains)) are configured as tcp proxy
+	// If mTLS permissive is enabled, there are five filter chains. The filter chain match should be
+	//  FCM 1: ALPN [istio-http/1.0, istio-http/1.1, istio-h2] Transport protocol: tls      --> HTTP traffic from sidecar over TLS
+	//  FCM 2: ALPN [http/1.0, http/1.1, h2c] Transport protocol: N/A                       --> HTTP traffic over plain text
+	//  FCM 3: ALPN [istio] Transport protocol: tls                                         --> TCP traffic from sidecar over TLS
+	//  FCM 4: ALPN [] Transport protocol: N/A                                              --> TCP traffic over plain text
+	//  FCM 5: ALPN [] Transport protocol: tls                                              --> TCP traffic over TLS
+	// If traffic is over plain text or mTLS is strict mode, there are two filter chains. The filter chaim match should be
+	//  FCM 1: ALPN [http/1.0, http/1.1, h2c, istio-http/1.0, istio-http/1.1, istio-h2]     --> HTTP traffic over plain text or TLS
+	//  FCM 2: ALPN []                                                                      --> TCP traffic over plain text or TLS
+	permissiveFilterChainMatchOptions = []FilterChainMatchOptions{
+		{
+			ApplicationProtocols: applicationProtocolsOverTLS,
+			TransportProtocol:    "tls",
+			Protocol:             plugin.ListenerProtocolHTTP,
+		},
+		{
+			ApplicationProtocols: applicationProtocols,
+			Protocol:             plugin.ListenerProtocolHTTP,
+		},
+		{
+			ApplicationProtocols: defaultApplicationProtocol,
+			TransportProtocol:    "tls",
+		},
+		{},
+		{
+			TransportProtocol: "tls",
```

Comment:
this is for inbound one-way TLS traffic into the sidecar->app ? [i.e. for ports marked as HTTPS or something?]

Reply:
It's for inbound TLS traffic from client without sidecar and we don't want to terminate the TLS connection.